File: Day250/removeDuplicatesFromString.go
package main 

import (
	"fmt"
)

func removeDuplicateLetters(s string) string {
	var answer []rune

	visited := make([]bool, 26)
	arr := make([]int, 20)

	for _,v := range s {
		arr[v - 'a']++
	}
	// fmt.Println(arr)

	stack := []rune{}

	for _, v := range s {
		if visited[v-'a'] {
			arr[v-'a']--
			continue
		}
		for len(stack) > 0 && stack[len(stack)-1] > v && arr[stack[len(stack)-1]-'a'] > 0 {
			visited[stack[len(stack)-1] - 'a'] = false
			stack = stack[:len(stack)- 1]
		}

		stack = append(stack, v)
		visited[v - 'a'] = true
		arr[v - 'a']--
	}
	// fmt.Println(visited)
	// fmt.Println(stack)

	for _, v := range stack {
		answer = append(answer, v)
	}

	return string(answer)
}

func main() {
	str1 :="bcabc" 
	str2 := "cbacdcbc"
	str3 := "cdadabcc"
	fmt.Println(removeDuplicateLetters(str1))
	fmt.Println(removeDuplicateLetters(str2))
	fmt.Println(removeDuplicateLetters(str3))
}

File: Day250/removeDuplicatesFromString.java
package Day250;

import java.util.Stack;

public class removeDuplicatesFromString {

    public static String removeDuplicateLetters(String s) {
        StringBuilder sb = new StringBuilder();

        boolean visited[] = new boolean[26];
        int arr[] = new int[26];
        
        for (int i=0; i<s.length(); i++) {
            arr[s.charAt(i)-'a']++;
        }

        Stack<Character> stack = new Stack<>();

        for (int i=0; i<s.length(); i++) {
            char temp = s.charAt(i);
            if (visited[temp - 'a']) {
                arr[temp - 'a']--;
                continue;
            }
            while (!stack.isEmpty() && stack.peek() > temp && arr[stack.peek() - 'a'] > 0) {
                visited[stack.peek() - 'a'] = false;
                stack.pop();
            }
            stack.push(temp);
            visited[temp - 'a'] = true;
            arr[temp - 'a']--;
        }

        // System.out.println(stack);

        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }

        return sb.reverse().toString();
    }

    public static void main(String[] args) {
        String str = "bcabc";
        String str2 = "cbacdcbc";
        String str3 = "cdadabcc";
        System.out.println(removeDuplicateLetters(str));
        System.out.println(removeDuplicateLetters(str2));
        System.out.println(removeDuplicateLetters(str3));
    }
}